File: backend/src/main/java/com/codeloom/backend/controller/TestCaseController.java
package com.codeloom.backend.controller;

import com.codeloom.backend.dto.CreateTestCaseRequest;
import com.codeloom.backend.dto.UpdateTestCaseRequest;
import com.codeloom.backend.model.TestCase;
import com.codeloom.backend.service.TestCaseService;
import jakarta.validation.Valid;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/v1/testCases")
@RequiredArgsConstructor
public class TestCaseController {
    private final TestCaseService service;

    @GetMapping("/{id}")
    public TestCase findById(@PathVariable UUID id) {
        return service.findById(id);
    }

    @GetMapping("/by-ids")
    public Iterable<TestCase> findAllByIds(@RequestParam List<UUID> ids) {
        return service.findAllByIds(ids);
    }

    @GetMapping("/by-problem-id/{problemId}")
    public Iterable<TestCase> findByProblemId(
            @PathVariable long problemId, @RequestParam(required = false) Boolean isPublic) {
        return service.findAllByProblemId(problemId, isPublic);
    }

    @PostMapping
    public TestCase create(@Valid @RequestBody CreateTestCaseRequest request) {
        return service.create(request);
    }

    @PutMapping("/{id}")
    public TestCase update(@PathVariable UUID id, @Valid @RequestBody UpdateTestCaseRequest request) {
        return service.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void delete(@PathVariable UUID id) {
        service.delete(id);
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/controller/TestCaseController.kt
package com.codeloom.backend.controller

import com.codeloom.backend.config.NOT_FOUND_RESPONSE_REF
import com.codeloom.backend.dto.CreateTestCaseRequest
import com.codeloom.backend.model.TestCase
import com.codeloom.backend.service.TestCaseService
import io.swagger.v3.oas.annotations.responses.ApiResponse
import jakarta.validation.Valid
import org.springframework.web.bind.annotation.*
import tools.jackson.databind.JsonNode
import java.util.*

@RestController
@RequestMapping("/v1/testCases")
class TestCaseController(private val testCaseService: TestCaseService) {
    @GetMapping("/{id}")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    fun findById(
        @PathVariable id: UUID,
    ): TestCase = testCaseService.findById(id)

    @GetMapping("/by-ids")
    fun findAllByIds(
        @RequestParam ids: List<UUID>,
    ): Iterable<TestCase> = testCaseService.findAllByIds(ids)

    @GetMapping("/by-problem-id/{problemId}")
    fun findByProblemId(
        @PathVariable problemId: Long,
        @RequestParam(required = false) isPublic: Boolean?,
    ): Iterable<TestCase> = testCaseService.findAllByProblemId(problemId, isPublic)

    @PostMapping
    fun create(
        @Valid @RequestBody request: CreateTestCaseRequest,
    ): TestCase = testCaseService.create(request)

    @PatchMapping("/{id}")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    fun patch(
        @PathVariable id: UUID,
        @RequestBody patchNode: JsonNode,
    ): TestCase = testCaseService.patch(id, patchNode)

    @DeleteMapping("/{id}")
    fun delete(
        @PathVariable id: UUID,
    ) = testCaseService.delete(id)
}
